bavi/modules/url.py

```diff
@@ -0,0 +1,109 @@
+import logging
+import requests
+from lxml import html
+from bs4 import UnicodeDammit
+from urllib.parse import urlparse
+import re
+
+log = logging.getLogger('bavi.modules.url')
+
+URL_REGEX = r'(https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=\u263a-\U0001f645]{1,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*))'
+
+IP_REGEX = r"(https?:\/\/([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(?<!172\.(16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31))(?<!127)(?<!10)(?<!0)\.([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(?<!192\.168)(?<!172\.(16|17|18|19|20|21|22|23|24|25|26|27|28|29|30|31))\.([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])(?<!\.255$)([-a-zA-Z0-9@:%_\+.~#?&//=]*))"
+
+IGNORE_REGEX = r'(?i)\.(jpg|jpg:large|png|gif|bmp|tiff|psd|mp4|mkv|avi|mov|mpg|vob|mp3|aac|wav|flac|ogg|mka|wma|zip|rar|7z|tar|iso|exe|dll|cfg|rc|reg|pxe|efi)$'
+
+compiled_url_regex = re.compile(URL_REGEX, re.UNICODE)
+compiled_ip_regex = re.compile(IP_REGEX)
+compiled_ignore_regex = re.compile(IGNORE_REGEX)
+
+TITLE_FORMAT = '[ {0} ] - {1}'
+
+
+def init(bot):
+    bot.add_command('title', title_command)
+    bot.add_matcher(compiled_url_regex, title_command)
+    bot.add_matcher(compiled_ip_regex, title_command)
+
+
+def title_command(bot, source, target, message, **kwargs):
+    log.debug('url module triggered on: ' + message)
+
+    if len(message) > 0:
+        valid_urls = get_valid_urls(message)
+
+        # make sure we have any good url, otherwise ignore
+        if (len(valid_urls) > 0):
+            for url in valid_urls:
+                response = get_user_readable_response(url)
+
+                if len(response) > 0:
+                    bot.say(target, response)
+
+
+def get_user_readable_response(url):
+    response = ''
+
+    if (len(url) > 0):
+        #now start a request for the page title
+        title = get_http_title(url)
+
+        if (len(title) > 0):
+            response = TITLE_FORMAT.format(title, get_domain_from_url(url))
+
+    return response
+
+
+def get_valid_urls(string_statement):
+    valid_matches = []
+    valid_urls = []
+
+    valid_matches += re.findall(compiled_url_regex, string_statement)
+    valid_matches += re.findall(compiled_ip_regex, string_statement)
+
+    log.debug(valid_matches)
+
+    for match in valid_matches:
+        if compiled_ignore_regex.search(match[0]) is None:
+            valid_urls.append(match[0])
+
+    log.debug('Found: ' + ', '.join(str(url) for url in valid_urls))
+
+    return valid_urls
+
+
+def get_http_title(url):
+    title = ''
+
+    try:
+        log.debug('Requesting url %s', url)
+
+        #TODO: bring these parameters to user defined setting.
+        #only request first 1024 bytes from server and timeout at 500ms
+        headers = {"Range": "bytes=0-1024"}
+        r = requests.get(url, headers=headers, timeout=0.5)
```

Comment:
500ms seems a little low.  Somewhere around 5 seconds feels like the sweet spot between waiting for slow sites to load and not dumping no-longer-relevant URL titles into chat.